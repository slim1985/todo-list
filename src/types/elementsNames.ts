export enum ElementsNames {
    STATUS_SELECT = 'status-select',
    TITLE_TEXT_AREA = 'title-textarea',
    DESCRIPTION_TEXT_AREA = 'description-textarea',
}
