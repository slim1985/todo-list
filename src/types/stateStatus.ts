export enum StateStatus {
    LOADING = 'loading',
    IDLE = 'idle',
    FAILED = 'failed',
}
