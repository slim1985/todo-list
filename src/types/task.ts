export interface Task {
    id: string;
    title: string;
    description: string;
    state: TaskStates;
}

export enum TaskStates {
    new = 'new',
    active = 'active',
    completed = 'completed',
}
