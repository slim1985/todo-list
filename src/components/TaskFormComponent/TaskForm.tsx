import React from 'react';
import { Task, TaskStates, TaskStateLabels } from '../../types/task';

enum ControlNames {
    Title = 'title',
    Description = 'description',
    Status = 'status',
}

export interface TaskFormProps {
    task: Task | null;
    hideTaskForm: () => void;
    createTask: (
        title: string,
        description: string,
        status: TaskStates,
    ) => void;
    updateTask: (
        id: string,
        title: string,
        description: string,
        status: TaskStates,
    ) => void;
    deleteTask: (taskId: string) => void;
}

export function TaskForm({
    task,
    hideTaskForm,
    createTask,
    updateTask,
    deleteTask,
}: TaskFormProps): JSX.Element {
    const [currentTask, setCurrentTask] = React.useState<Task>({
        id: task?.id ?? '',
        status: task?.status ?? TaskStates.NEW,
        title: task?.title ?? 'Title',
        description: task?.description ?? 'Description',
    });

    function onTaskChange(
        event:
            | React.ChangeEvent<HTMLSelectElement>
            | React.ChangeEvent<HTMLTextAreaElement>,
    ): void {
        const { name, value } = event.target;

        setCurrentTask({
            ...currentTask,
            [name]: value,
        });
    }

    function onSaveClick(): void {
        if (!currentTask.id) {
            createTask(
                currentTask.title,
                currentTask.description,
                currentTask.status,
            );
        } else {
            updateTask(
                currentTask.id,
                currentTask.title,
                currentTask.description,
                currentTask.status,
            );
        }

        hideTaskForm();
    }

    function onDeleteTaskClick(): void {
        deleteTask(currentTask.id);
        hideTaskForm();
    }

    return (
        <div className="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]">
            <div className="task-form flex flex-col flex-nowrap bg-white bg-gray-300 rounded-md">
                <select
                    className="text-3xl m-1"
                    name={ControlNames.Status}
                    value={currentTask.status}
                    onChange={(e) => onTaskChange(e)}
                >
                    {Object.keys(TaskStates).map((state, index) => (
                        <option key={index} value={index}>
                            {TaskStateLabels[Object.values(TaskStates)[index]]}
                        </option>
                    ))}
                </select>
                <textarea
                    className="h-24 border-solid border-2 border-yellow-200 rounded m-1 p-1"
                    name={ControlNames.Title}
                    value={currentTask.title}
                    onChange={(e) => onTaskChange(e)}
                ></textarea>
                <textarea
                    className="h-full border-solid border-2 border-yellow-200 rounded m-1 p-1"
                    name={ControlNames.Description}
                    value={currentTask.description}
                    onChange={(e) => onTaskChange(e)}
                ></textarea>
                <div className="flex justify-center my-5 mx-3 space-x-7">
                    <button
                        onClick={() => onSaveClick()}
                        className="size-20 p-1 bg-green-300 rounded-md"
                    >
                        Save
                    </button>
                    <button
                        onClick={() => hideTaskForm()}
                        className="size-20 p-1 bg-yellow-300 rounded-md"
                    >
                        Cancel
                    </button>
                    <button
                        onClick={() => onDeleteTaskClick()}
                        className="size-20 p-1 bg-red-300 rounded-md"
                    >
                        Delete
                    </button>
                </div>
            </div>
        </div>
    );
}
