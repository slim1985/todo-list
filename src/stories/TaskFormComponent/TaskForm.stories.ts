import type { Meta, StoryObj } from '@storybook/react';
import { TaskForm } from '../../components/TaskFormComponent/TaskForm';
import { Task, TaskStates } from '../../types/task';
import tasks from '../../mocks/data-mock.json';

type Story = StoryObj<typeof meta>;

const meta: Meta<typeof TaskForm> = {
    component: TaskForm,
    parameters: {
        viewport: {
            defaultViewport: 'mobile1',
        },
    },
};

const taskList = JSON.parse(JSON.stringify(tasks)) as Task[];

export default meta;

export const TaskFormEmpty: Story = {
    name: 'Empty',
    args: {
        task: {
            id: '',
            title: '',
            description: '',
            status: TaskStates.NEW,
        },
    },
};

export const TaskFormPartialFilled: Story = {
    name: 'Partial Filled',
    args: {
        task: taskList[1],
    },
};

export const TaskFormFullFilled: Story = {
    name: 'Full Filled',
    args: {
        task: {
            ...taskList[0],
            description:
                'lorem ipsum dolor sit amet consectetur adipiscing elit sed do eiusmod tempor incididunt ut labore et dolore magna aliqua ut enim ad minim veniam quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat duis aute irure dolor in reprehenderit voluptate velit esse cillum dolore fugiat nulla pariatur excepteur sint occaecat cupidatat non proident sunt culpa qui officia deserunt mollit anim id est laborum',
            status: TaskStates.COMPLETED,
        },
    },
};
