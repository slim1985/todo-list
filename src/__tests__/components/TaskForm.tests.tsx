import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { TaskForm } from '../../components/TaskFormComponent/TaskForm';
import { Task, TaskStateLabels, TaskStates } from '../../types/task';
import { StateStatus } from '../../types/stateStatus';

const task: Task = {
    id: '1',
    title: 'Task 1',
    description: 'Task 1 description',
    status: TaskStates.NEW,
};

const stateStatus = StateStatus.IDLE;
const hideTaskForm = jest.fn();
const createTask = jest.fn();
const updateTask = jest.fn();
const deleteTask = jest.fn();

describe('TaskForm', () => {
    it('TaskForm renders correctly with idle state', () => {
        // Arrange.
        // Act.
        const { asFragment } = render(
            <TaskForm
                task={task}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('TaskForm renders correctly with loading state', () => {
        // Arrange.
        // Act.
        const { asFragment } = render(
            <TaskForm
                task={task}
                stateStatus={StateStatus.LOADING}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('TaskForm renders correctly with failed state', () => {
        // Arrange.
        // Act.
        const { asFragment } = render(
            <TaskForm
                task={task}
                stateStatus={StateStatus.FAILED}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('TaskForm data renders correctly', () => {
        // Arrange.
        // Act.
        const { getByText } = render(
            <TaskForm
                task={task}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        // Assert.
        expect(getByText(task.title)).toBeInTheDocument();
        expect(getByText(task.description)).toBeInTheDocument();
        expect(getByText(TaskStateLabels[task.status])).toBeInTheDocument();
    });

    it('TaskForm Save calls createTask', () => {
        // Arrange.
        // Act.
        render(
            <TaskForm
                task={null}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        screen.getByText('Save').click();

        // Assert.
        expect(createTask).toHaveBeenCalledTimes(1);
    });

    it('TaskForm Save calls updateTask', () => {
        // Arrange.
        // Act.
        render(
            <TaskForm
                task={task}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        screen.getByText('Save').click();

        // Assert.
        expect(updateTask).toHaveBeenCalledTimes(1);
    });

    it('TaskForm Delete calls deleteTask', () => {
        // Arrange.
        // Act.
        render(
            <TaskForm
                task={task}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        screen.getByText('Delete').click();

        // Assert.
        expect(deleteTask).toHaveBeenCalledTimes(1);
    });

    it('TaskForm Cancel calls hideTaskForm', () => {
        // Arrange.
        // Act.
        render(
            <TaskForm
                task={task}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        screen.getByText('Cancel').click();

        // Assert.
        expect(hideTaskForm).toHaveBeenCalledTimes(1);
    });

    it('TaskForm onTaskChange updates currentTask', () => {
        // Arrange.
        const newTitle = 'New Title';
        const newDescription = 'New Description';

        // Act.
        render(
            <TaskForm
                task={task}
                stateStatus={stateStatus}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
            />,
        );

        fireEvent.change(screen.getByText(task.title), {
            target: { name: 'title', value: newTitle },
        });
        fireEvent.change(screen.getByText(task.description), {
            target: { name: 'description', value: newDescription },
        });
        fireEvent.change(screen.getByRole('combobox'), {
            target: { name: 'status', value: TaskStates.COMPLETED },
        });

        // Assert.
        expect(screen.getByText(newTitle)).toBeInTheDocument();
        expect(screen.getByText(newTitle)).toBeInTheDocument();
        expect(
            screen.getByText(TaskStateLabels[TaskStates.COMPLETED]),
        ).toBeInTheDocument();
    });
});
