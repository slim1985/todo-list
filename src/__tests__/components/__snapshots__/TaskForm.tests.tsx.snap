// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm TaskForm renders correctly with failed state 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]"
  >
    <div
      class="flex flex-col flex-nowrap w-5/6 h-5/6 bg-white bg-gray-300 rounded-md retina:rounded-3xl"
    >
      <div
        class="text-center text-3xl m-1 text-red-500"
      >
        Failed to save task. Check internet connection and try it again.
      </div>
      <select
        class="text-3xl retina:text-8xl m-1 retina:m-3"
        name="status"
      >
        <option
          value="0"
        >
          New
        </option>
        <option
          value="1"
        >
          Active
        </option>
        <option
          value="2"
        >
          Completed
        </option>
      </select>
      <textarea
        class="h-24 retina:h-64 border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
        name="title"
      >
        Task 1
      </textarea>
      <textarea
        class="h-full border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
        name="description"
      >
        Task 1 description
      </textarea>
      <div
        class="flex justify-center my-5 retina:my-5 mx-3 retina:mx-8 space-x-7"
      >
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-green-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Save
        </button>
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-yellow-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Cancel
        </button>
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-red-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TaskForm TaskForm renders correctly with idle state 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]"
  >
    <div
      class="flex flex-col flex-nowrap w-5/6 h-5/6 bg-white bg-gray-300 rounded-md retina:rounded-3xl"
    >
      <select
        class="text-3xl retina:text-8xl m-1 retina:m-3"
        name="status"
      >
        <option
          value="0"
        >
          New
        </option>
        <option
          value="1"
        >
          Active
        </option>
        <option
          value="2"
        >
          Completed
        </option>
      </select>
      <textarea
        class="h-24 retina:h-64 border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
        name="title"
      >
        Task 1
      </textarea>
      <textarea
        class="h-full border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
        name="description"
      >
        Task 1 description
      </textarea>
      <div
        class="flex justify-center my-5 retina:my-5 mx-3 retina:mx-8 space-x-7"
      >
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-green-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Save
        </button>
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-yellow-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Cancel
        </button>
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-red-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TaskForm TaskForm renders correctly with loading state 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]"
  >
    <div
      class="flex justify-center items-center z-50 fixed h-full w-full overflow-hidden bg-black/[.5]"
    >
      <div
        role="status"
      >
        <svg
          aria-hidden="true"
          class="w-24 retina:w-56 h-24 retina:h-60 text-gray-200 animate-spin dark:text-gray-600 fill-blue-600"
          fill="none"
          viewBox="0 0 100 101"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z"
            fill="currentColor"
          />
          <path
            d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z"
            fill="currentFill"
          />
        </svg>
        <span
          class="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
    <div
      class="flex flex-col flex-nowrap w-5/6 h-5/6 bg-white bg-gray-300 rounded-md retina:rounded-3xl"
    >
      <select
        class="text-3xl retina:text-8xl m-1 retina:m-3"
        name="status"
      >
        <option
          value="0"
        >
          New
        </option>
        <option
          value="1"
        >
          Active
        </option>
        <option
          value="2"
        >
          Completed
        </option>
      </select>
      <textarea
        class="h-24 retina:h-64 border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
        name="title"
      >
        Task 1
      </textarea>
      <textarea
        class="h-full border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
        name="description"
      >
        Task 1 description
      </textarea>
      <div
        class="flex justify-center my-5 retina:my-5 mx-3 retina:mx-8 space-x-7"
      >
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-green-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Save
        </button>
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-yellow-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Cancel
        </button>
        <button
          class="size-20 retina:size-60 p-1 retina:p-3 bg-red-300 rounded-md retina:rounded-3xl retina:text-6xl"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
